frontend/src/metabase/scaffolding/utils.ts

```diff
@@ -128,53 +139,249 @@ export function renderValue(
+    return null;
+  }
+  return names.reduce((acc: Field[], name: string) => {
+    acc.push(fieldsByName[name]);
+    return acc;
+  }, []);
+}
+
+function deleteFields(
+  fieldsByName: Record<string, Field>,
+  fields: Field[],
+): void {
+  fields.forEach((field) => {
+    delete fieldsByName[field.name];
+  });
+}
+
+function sectionFields(fields: Field[]): { field_id: number }[] {
+  return fields.map((field) => ({
+    field_id: getRawTableFieldId(field),
+  }));
+}
+
 export function getDefaultObjectViewSettings(
   table: Table | undefined,
 ): ObjectViewSettings {
   const fields = table?.fields ?? [];
-  const headerFields = fields
-    .filter((f) => isEntityName(f) || isPK(f))
-    .slice(0, 3);
-  // const normalFields = fields.filter((f) => !headerFields.includes(f));
 
-  return {
-    sections: [
-      {
+  const fieldsByName = fields.reduce<Record<string, Field>>((acc, field) => {
+    acc[field.name] = field;
+    return acc;
+  }, {});
+
+  const leftoverFields = JSON.parse(JSON.stringify(fieldsByName));
+
+  const pkFields = fields.filter((f) => isPK(f));
+  deleteFields(leftoverFields, pkFields);
+
+  // Find things that look like part of a name.
+  let nameFields: Field[] =
+    notEmpty(fields.filter((f) => isEntityName(f))) ||
+    fieldsIfPresent(fieldsByName, "name") ||
+    fieldsIfPresent(fieldsByName, "title", "full_name") ||
+    fieldsIfPresent(fieldsByName, "full_name") ||
+    fieldsIfPresent(fieldsByName, "title", "first_name", "last_name") ||
+    fieldsIfPresent(fieldsByName, "first_name", "last_name") ||
+    // Or, failing that, a title.
+    notEmpty(fields.filter((f) => isTitle(f))) ||
+    fieldsIfPresent(fieldsByName, "title") ||
+    [];
+
+  // If there's only promising option, use that.
+  if (nameFields.length === 0) {
+    const nonPkFields = fields.filter((f) => !isPK(f));
+    if (nonPkFields.length === 1) {
+      nameFields = nonPkFields;
+    } else {
+      const stringFields = nonPkFields.filter(
+        (f) => !isPK(f) && f.base_type === "type/Text",
+      );
+      if (stringFields.length === 1) {
+        nameFields = stringFields;
+      }
+    }
+  }
+
+  // If there's only one kind of prefixed name, let's go with that.
+  if (nameFields.length === 0) {
+    const somethingNameFields = fields.filter((f) =>
+      f.name.toLowerCase().endsWith("name"),
+    );
+    if (somethingNameFields.length === 1) {
+      nameFields = somethingNameFields;
+    }
+  }
+  deleteFields(leftoverFields, nameFields);
+
+  // Is there any overlap between the primary key and the name fields?
+  // This is O(n^2) so sue me.
+  const pkInName = pkFields.some((f) =>
+    nameFields.some((nf) => nf.name === f.name),
+  );
+
+  const headerFields =
+    !pkInName && pkFields.length > 0 && pkFields.length + nameFields.length <= 3
+      ? [...pkFields, ...nameFields]
+      : // if you have more than 3 entity name or title fields, we'll use them *all*
+        nameFields;
```

Comment:
We need to respect the limit of max 3 fields in header

Reply:
Is that a hard display limit?

As a user I would think it's broken to split these explicit title fields up.

Reply:
Guess I will just truncate for now, and we can revisit.